core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -536,10 +541,18 @@ class ReplicaManager(val config: KafkaConfig,
            * This can cause a replica to always be out of sync.
            */
           val initialLogEndOffset = localReplica.logEndOffset
+
           val logReadInfo = localReplica.log match {
             case Some(log) =>
               val adjustedFetchSize = if (Topic.isInternal(topic) && !readOnlyCommitted) Math.max(fetchSize, log.config.maxMessageSize) else fetchSize
-              log.read(offset, adjustedFetchSize, maxOffsetOpt)
+
+              //Try the read first, this tells us whether we need all of adjustedFetchSize for this partition
+              var fetch = log.read(offset, adjustedFetchSize, maxOffsetOpt)
+
+              //If the partition is marked as throttled, and we are over-quota then exclude it
+              if (quota.isThrottled(TopicAndPartition(topic, partition)) && quota.isQuotaExceeded) //TODO - could also check ISR inclusion??
```

Comment:
With this, after recording the bytes, the quota could be exceeded? The earlier approach where we use  quota.isQuotaExceed(expectedBytes) seems more conservative and is less likely for quota to be exceeded. Is there a reason not to use quota.isQuotaExceed(expectedBytes)?